Automattic/jetpack
# Boost: Refactor and add new e2e tests

Fixes #21756

#### Changes proposed in this Pull Request:


This PR implements the core E2E tests described in #21756. It is important to note there is not exactly a 1-1 relationship between the tests in this PR and the tests scenarios highlighted in #21756 but the tests in this PR will cover scenarios by doing a combination of steps and checks.

It adds some new Jetpack Boost WP CLI commands to allow managing tests prerequisites more easily. These commands
- A command to manage Jetpack Boost connection to WPCOM. it can be used to activate/deactivate and check the connection status.
- A command to clear any jetpack Boost cache, posts and related options to be able to bootstrap a clean Boost environment.

It creates a new prerequisite allowing setting up some test posts useful to test the Lazy Images and Defer JS modules. These test post pages can be accessed during a test via a new `projects/plugins/boost/tests/e2e/lib/pages/TestContentPage.js`. The appropriate test post page is retrieved using its id number which is retrieved using the post title. This is to avoid messing out and changing the permalinks structure since the repo files and therefore the `.htaccess` file are shared between the local development environment and the test environment. This might not be ideal but it seems to work pretty well.

This PR also make some refactoring and additions to the E2E Common files package to support the Jetpack Boost E2E tests. I tried to make them generic so they could be re-used in other plugins. These changes are:
- Adding the new `selectJetpackBoost` method to `tools/e2e-commons/pages/wp-admin/sidebar.js` to allow accessing the Jetpack Boost settings page. It could be contained inside the jetpack Boost E2E files but I thought it makes sense to keep all sidebar actions in that file.
- Refactoring the methods to activate and deactivate the Jetpack plugin from the Plugins page (using actions such as clicking on activate/deactivate links) in `tools/e2e-commons/pages/wp-admin/plugins.js` to make them more generic so they can be used to activate and inactivate any plugins.
- Add prerequisite actions to the `tools/e2e-commons/env/prerequisites.js` files to allow activating and deactivating plugins using the WP-CLI commands.

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
N/A

#### Testing instructions:
- The Jetpack Boost E2E GitHub action should pass.


Locally it can be tested doing the following:

- run `pnpm install` at the root of the repo
-  run `jetpack install plugins/jetpack`
- run `jetpack install plugins/boost`
- in `projects/plugins/boost/tests/e2e`  run `pnpm install`
- then in that same folder run `pnpm run test-decrypt-config`
- run `pnpm run env-start && pnpm run tunnel-on`
- and finally `run pnpm run test-e2e`

Also individual specs file can be run using for the command `pnpm run test-e2e -- path/to/the/spec/file` for example `pnpm run test-e2e -- specs/speed-score.test.j`.

Also the debug mode can be activated using the following command `PWDEBUG=1 pnpm run test-e2e -- path/to/the/spec/file` or `PWDEBUG=1 pnpm run test-e2e` for the whole test suite. The `PWDEBUG=1` statement makes sure Chromium becomes visible so you can see the actual tests being executed. You will have to manually move some tests step forward from the Playwright Inspector window. See the video below:


https://user-images.githubusercontent.com/927932/145340967-13482b2c-c201-4284-bb82-5313a42e1e3b.mp4

Finally, you once you did the original `pnpm run env-start && pnpm run tunnel-on` command, you can access your test site from any browser using the tunnel URL provided by the ` pnpm run tunnel-on` command.
